src/levanter/models/attention.py

```diff
@@ -57,27 +59,77 @@ def dot_product_attention(
         raise ValueError("QPos and KPos must be different")
 
     if use_flash:
-        from levanter.models.flash_attention import BLOCK_SIZE, flash_attention
-
-        if flash_block_size is None:
-            flash_block_size = BLOCK_SIZE
-
-        return flash_attention(
-            QPos,
-            KPos,
-            Key,
-            query,
-            key,
-            value,
-            block_size=flash_block_size,
-            mask=mask,
-            bias=bias,
-            dropout=dropout,
-            inference=inference,
-            key=prng,
-            dtype=attention_dtype,
-            precision=precision,
-        )
+        accelerator_type = xla_bridge.get_backend().platform.lower()
+
+        if accelerator_type == "tpu" or accelerator_type == "cpu":
+            from levanter.models.flash_attention import flash_attention
+
+            if flash_block_size is None:
+                seq_len = query.shape["position"]
+                flash_block_size = seq_len
+
+            return flash_attention(
+                QPos,
+                KPos,
+                Key,
+                query,
+                key,
+                value,
+                block_size=flash_block_size,
+                mask=mask,
+                bias=bias,
+                dropout=dropout,
+                inference=inference,
+                key=prng,
+                dtype=attention_dtype,
+                precision=precision,
+            )
+        elif accelerator_type == "gpu":
+            from transformer_engine.jax.fused_attn import AttnBiasType, AttnMaskType, self_fused_attn
+
+            # TransformerEngine looks for qkv shape:
+            # batch_shape, max_seqlen, nqkv, num_heads, head_dim; where nqkv = 3
+            q_ = haliax.rearrange(query, ("batch", "position", "heads", "head_size")).array
+            k_ = haliax.rearrange(key, ("batch", "key_position", "heads", "head_size")).array
+            v_ = haliax.rearrange(value, ("batch", "key_position", "heads", "head_size")).array
+
```

Comment:
I'd prefer these not be hard coded. At a minimum, could we do something like:
 
```suggestion
            
            q_ = haliax.rearrange(query, ("batch", QPos, "heads", Key)).array
            k_ = haliax.rearrange(key, ("batch", KPos, "heads", Key)).array
            v_ = haliax.rearrange(value, ("batch", KPos, "heads", Key)).array
```

Ideally we wouldn't hardcode the batch and heads dim either, but we'll need a heuristic for that. (If it were just one batch-y dim we could flatten them all)

Reply:
probably the thing to do is to choose the latest non-pos/non-key dim as a heads dim and then convert the rest to batch dims, but let's not worry about that right now